File: Corte 2/Ejericios/ordenamiento.cpp
#include <iostream>
#include <chrono>   // para medir el tiempo
using namespace std;

/* Bubble Sort */
void bubbleSort(int arr[], int n, long long &comparaciones, long long &intercambios) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            comparaciones++;                     // se compara arr[j] con arr[j+1]
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                intercambios++;                  // hubo intercambio
            }
        }
    }
}
/* Selection Sort */
void selectionSort(int arr[], int n, long long &comparaciones, long long &intercambios) {
    for (int i = 0; i < n - 1; i++) {
        int minIdx = i;
        for (int j = i + 1; j < n; j++) {
            comparaciones++;                     // se compara arr[j] con arr[minIdx]
            if (arr[j] < arr[minIdx]) {
                minIdx = j;
            }
        }
        int temp = arr[minIdx];
        arr[minIdx] = arr[i];
        arr[i] = temp;
        intercambios++;                          // un intercambio por cada pasada
    }
}
/* Insertion Sort */
void insertionSort(int arr[], int n, long long &comparaciones, long long &intercambios) {
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        // La comparacion vivia dentro del while; se separa para poder contarla.
        while (j >= 0) {
            comparaciones++;                     // se compara arr[j] con key
            if (arr[j] <= key) break;            // misma condicion de corte que el while original
            arr[j + 1] = arr[j];
            intercambios++;                      // desplazamiento (movimiento de dato)
            j--;
        }
        arr[j + 1] = key;
    }
}

/* ---------- Medicion de tiempo, comparaciones e intercambios ---------- */
int main() {
    const int n = 10000;   // tamano del arreglo de prueba

    // arreglo base con numeros en orden inverso (peor caso aproximado)
    int base[n];
    for (int i = 0; i < n; i++) base[i] = n - i;

    // copias para que cada algoritmo ordene los MISMOS datos
    int a[n], b[n], c[n];
    for (int i = 0; i < n; i++) { a[i] = base[i]; b[i] = base[i]; c[i] = base[i]; }

    long long comparaciones, intercambios;
    auto inicio = chrono::high_resolution_clock::now();
    auto fin = inicio;

    // --- Bubble Sort ---
    comparaciones = 0; intercambios = 0;
    inicio = chrono::high_resolution_clock::now();
    bubbleSort(a, n, comparaciones, intercambios);
    fin = chrono::high_resolution_clock::now();
    double tBubble = chrono::duration<double, milli>(fin - inicio).count();
    cout << "Bubble Sort    -> tiempo: " << tBubble << " ms"
         << " | comparaciones: " << comparaciones
         << " | intercambios: " << intercambios << endl;

    // --- Selection Sort ---
    comparaciones = 0; intercambios = 0;
    inicio = chrono::high_resolution_clock::now();
    selectionSort(b, n, comparaciones, intercambios);
    fin = chrono::high_resolution_clock::now();
    double tSelection = chrono::duration<double, milli>(fin - inicio).count();
    cout << "Selection Sort -> tiempo: " << tSelection << " ms"
         << " | comparaciones: " << comparaciones
         << " | intercambios: " << intercambios << endl;

    // --- Insertion Sort ---
    comparaciones = 0; intercambios = 0;
    inicio = chrono::high_resolution_clock::now();
    insertionSort(c, n, comparaciones, intercambios);
    fin = chrono::high_resolution_clock::now();
    double tInsertion = chrono::duration<double, milli>(fin - inicio).count();
    cout << "Insertion Sort -> tiempo: " << tInsertion << " ms"
         << " | comparaciones: " << comparaciones
         << " | intercambios: " << intercambios << endl;

    return 0;
}

File: Corte 2/Ejericios/ordenamiento.py
# comparo dos vecinos si estan al reves los intercambio y repito el proceso 
# hast que nadie mas se mueva bubble sort
#neccsito probar todos los tipos de ordenamieto con el siguiente arreglo
import time

datos = [64, 25, 12, 22, 11, 90, 45, 33]

def medir_tiempo(metodo, arr):
    inicio = time.perf_counter()
    # Usamos .copy() para no modificar la lista 'datos' original en cada prueba
    resultado, comparaciones, intercambios = metodo(arr.copy())
    fin = time.perf_counter()
    tiempo_total = fin - inicio
    return resultado, tiempo_total, comparaciones, intercambios

def bubble_sort(arr):
    n = len(arr)
    comparaciones = 0
    intercambios = 0
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            comparaciones += 1
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                intercambios += 1
                swapped = True
        if not swapped:
            break
    return arr, comparaciones, intercambios

def selection_sort(arr):    
    n = len(arr)
    comparaciones = 0
    intercambios = 0
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            comparaciones += 1
            if arr[j] < arr[min_idx]:
                min_idx = j
        if min_idx != i:
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
            intercambios += 1
    return arr, comparaciones, intercambios

#buscar el más pequeño y ponerlo al 
# principio selection sort

def insertion_sort(arr):
    comparaciones = 0
    intercambios = 0
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            comparaciones += 1
            arr[j + 1] = arr[j]
            j -= 1
        if j >= 0:
            comparaciones += 1
        arr[j + 1] = key
        intercambios += 1
    return arr, comparaciones, intercambios

# tomo el segubndo y lo inserto donde va 
# respecto al primero insertion sort

print("Bubble Sort:")
sorted_arr, tiempo, comparaciones, intercambios = medir_tiempo(bubble_sort, datos)
print(f"Comparaciones: {comparaciones}, Intercambios: {intercambios}, Tiempo: {tiempo:.7f} segundos\n")

print("Selection Sort:")
sorted_arr, tiempo, comparaciones, intercambios = medir_tiempo(selection_sort, datos)
print(f"Comparaciones: {comparaciones}, Intercambios: {intercambios}, Tiempo: {tiempo:.7f} segundos\n")

print("Insertion Sort:")
sorted_arr, tiempo, comparaciones, intercambios = medir_tiempo(insertion_sort, datos)
print(f"Comparaciones: {comparaciones}, Intercambios: {intercambios}, Tiempo: {tiempo:.7f} segundos\n")
